core/src/main/java/org/infinispan/reactive/publisher/impl/PartitionAwareClusterPublisherManager.java

```diff
@@ -108,48 +107,53 @@ public void start() {
-   public <R> SegmentCompletionPublisher<R> entryPublisher(IntSet segments, Set<K> keysToInclude,
+   public <R> SegmentPublisherSupplier<R> entryPublisher(IntSet segments, Set<K> keysToInclude,
          InvocationContext invocationContext, boolean includeLoader, DeliveryGuarantee deliveryGuarantee, int batchSize,
          Function<? super Publisher<CacheEntry<K, V>>, ? extends Publisher<R>> transformer) {
       checkPartitionStatus();
-      SegmentCompletionPublisher<R> original = super.entryPublisher(segments, keysToInclude, invocationContext,
+      SegmentPublisherSupplier<R> original = super.entryPublisher(segments, keysToInclude, invocationContext,
             includeLoader, deliveryGuarantee, batchSize, transformer);
       return registerPublisher(original);
    }
 
-   private <R> SegmentCompletionPublisher<R> registerPublisher(SegmentCompletionPublisher<R> original) {
-      return (subscriber, segmentsComplete) -> {
-         // Processor has to be serialized due to possibly invoking onError from a different thread
-         FlowableProcessor<R> earlyTerminatingProcessor = AsyncProcessor.<R>create().toSerialized();
-         pendingProcessors.add(earlyTerminatingProcessor);
-
-         // Have to check after registering in case if we got a partition between when the publisher was created and
-         // subscribed to
-         if (isPartitionDegraded()) {
-            pendingProcessors.remove(earlyTerminatingProcessor);
-            earlyTerminatingProcessor.onError(CLUSTER.partitionDegraded());
-            original.subscribe(earlyTerminatingProcessor);
-         } else {
-            Flowable<R> actualPublisher = Flowable.<R>fromPublisher(s -> original.subscribe(s, segmentsComplete))
-                  .doFinally(earlyTerminatingProcessor::onComplete);
-            Publisher<R> flowableAsPublisher = earlyTerminatingProcessor
-                  .doFinally(() -> pendingProcessors.remove(earlyTerminatingProcessor));
-            Flowable.merge(flowableAsPublisher, actualPublisher)
-                  .subscribe(subscriber);
+   private <R> SegmentPublisherSupplier<R> registerPublisher(SegmentPublisherSupplier<R> original) {
+      return new SegmentPublisherSupplier<R>() {
+         @Override
+         public Publisher<Notification<R>> publisherWithSegments() {
+            return handleEarlyTermination(SegmentPublisherSupplier::publisherWithSegments);
+         }
+
+         @Override
+         public Publisher<R> publisherWithoutSegments() {
+            return handleEarlyTermination(SegmentPublisherSupplier::publisherWithoutSegments);
+         }
+
+         private <S> Flowable<S> handleEarlyTermination(Function<SegmentPublisherSupplier<R>, Publisher<S>> function) {
+            CompletableFuture<Void> cf = new CompletableFuture<>();
+            pendingCompletableFutures.add(cf);
+
+            return Flowable.fromPublisher(function.apply(original))
+                           .doOnNext(s -> throwExceptionIfNeeded(cf))
```

Comment:
I believe this still has the issue that if a partition happens and no additional values are returned that this will not error. I am thinking we should probably add a .doOnComplete check as well.